### PR TITLE
Migrate scheduler api types to sized integers

### LINKED ISSUE #82138: Types in versioned scheduler api are not sized
/sig scheduling
/sig api-machinery

#53850 changed a some types to sized integers in https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/api/types.go but not in https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/api/v1/types.go

This leaves to problems when encoding and decoding between types. For example, this is blocking #81911

We should migrate all types to sized integers in both files.

### PR DESCRIPTION
**Pull request by ahmad-diaa**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Migrating `int` types to fixed sizes in https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/api/types.go and https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/api/v1/types.go. Having different integer fields led to difficulty in encoding and decoding between both types.

**Which issue(s) this PR fixes**:

Fixes #82138

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/scheduler/api/v1/types.go [hunk 1]
```diff
@@ -137,17 +137,17 @@ type RequestedToCapacityRatioArguments struct {
 // UtilizationShapePoint represents single point of priority function shape.
 type UtilizationShapePoint struct {
 	// Utilization (x axis). Valid values are 0 to 100. Fully utilized node maps to 100.
-	Utilization int `json:"utilization"`
+	Utilization int8 `json:"utilization"`
```
Comment (draveness) [inline pkg/scheduler/api/v1/types.go]: I don't think we should change it to int8 which is not common in the k/k

```
$ ag int8 pkg
pkg/controller/daemon/daemon_controller_test.go
580:			for pass := uint8(0); expectedLimit <= podControl.FakePodControl.CreateLimit; pass++ {

pkg/controller/job/job_controller_test.go
360:		for pass := uint8(0); expectedLimit <= tc.podLimit; pass++ {

pkg/controller/replicaset/replica_set_test.go
269:	for pass := uint8(0); expectedLimit <= fakePodControl.CreateLimit; pass++ {

pkg/kubelet/apis/podresources/v1alpha1/api.pb.go
398:				dAtA[i] = uint8(uint64(l)&0x7f | 0x80)
402:			dAtA[i] = uint8(l)
412:		dAtA[offset] = uint8(v&0x7f | 0x80)
416:	dAtA[offset] = uint8(v)

pkg/kubelet/apis/pluginregistration/v1/api.pb.go
309:				dAtA[i] = uint8(uint64(l)&0x7f | 0x80)
313:			dAtA[i] = uint8(l)
393:		dAtA[offset] = uint8(v&0x7f | 0x80)
397:	dAtA[offset] = uint8(v)

pkg/kubelet/apis/pluginregistration/v1alpha1/api.pb.go
560:		dAtA[offset] = uint8(v&0x7f | 0x80)
564:	dAtA[offset] = uint8(v)

pkg/kubelet/apis/pluginregistration/v1beta1/api.pb.go
560:		dAtA[offset] = uint8(v&0x7f | 0x80)
564:	dAtA[offset] = uint8(v)

pkg/kubelet/apis/deviceplugin/v1alpha/api.pb.go
1169:		dAtA[offset] = uint8(v&0x7f | 0x80)
1173:	dAtA[offset] = uint8(v)

pkg/kubelet/apis/deviceplugin/v1beta1/api.pb.go
1857:		dAtA[offset] = uint8(v&0x7f | 0x80)
1861:	dAtA[offset] = uint8(v)

pkg/kubelet/pluginmanager/pluginwatcher/example_plugin_apis/v1beta2/api.pb.go
324:		dAtA[offset] = uint8(v&0x7f | 0x80)
328:	dAtA[offset] = uint8(v)

pkg/kubelet/pluginmanager/pluginwatcher/example_plugin_apis/v1beta1/api.pb.go
323:		dAtA[offset] = uint8(v&0x7f | 0x80)
327:	dAtA[offset] = uint8(v)

pkg/registry/core/service/allocator/utils.go
31:var bitCounts = []int8{

pkg/kubectl/cmd/get/sorter.go
174:	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
176:	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
211:		case uint8:
212:			if jtype, ok := j.Interface().(uint8); ok {
227:		case int8:
228:			if jtype, ok := j.Interface().(int8); ok {
```

**Review by draveness — CHANGES_REQUESTED** [review]
/ok-to-test

File: pkg/scheduler/factory/plugins_test.go [hunk 2]
```diff
@@ -51,22 +51,12 @@ func TestAlgorithmNameValidation(t *testing.T) {
...
 }
 
-func TestValidatePriorityConfigOverFlow(t *testing.T) {
+func TestValidatePriorityConfig(t *testing.T) {
 	tests := []struct {
 		description string
 		configs     []priorities.PriorityConfig
 		expected    bool
 	}{
-		{
```
Comment (ahmad-diaa) [inline pkg/scheduler/factory/plugins_test.go]: The following tests doesn't seem relevant anymore for me, as the weight size is `int32` and the total Node Priority score is `int`. So, they are only relevant in case of 32-bit architectures.  But when it comes to 64-bit architectures it is really hard to overflow having a weight of size `32 bit` and a total priority score of size `64 bit`. Any thoughts on this? Is it fine to just remove those cases? Is there some other test cases that would be relevant and could be added (I couldn't come up with any)?

File: pkg/scheduler/api/v1/types.go (hunk 1 above)
Comment (ahmad-diaa) [reply pkg/scheduler/api/v1/types.go]: Do you think it's better to use int32 like other fields even if it's bounded by (0-100)?

File: pkg/scheduler/factory/plugins_test.go (hunk 2 above)
Comment (alculquicondor) [reply pkg/scheduler/factory/plugins_test.go]: considering that we move to int64, the tests are still valid. I think the second one is valid regardless.

File: pkg/scheduler/api/types.go [hunk 3]
```diff
@@ -34,7 +35,7 @@ const (
 	// MaxPriority defines the max priority value.
 	MaxPriority = 10
 	// MaxWeight defines the max weight value.
-	MaxWeight = MaxInt / MaxPriority
+	MaxWeight = int32(math.MaxInt32 / MaxPriority)
```
[the remark this answers is not in the document]
Comment (ahmad-diaa) [reply pkg/scheduler/api/types.go]: Maybe that's because MaxInt is not exported by the `math` package and it's used by `plugins_test.go`. Is that what you were referring to?

File: pkg/scheduler/api/types.go [hunk 4]
```diff
@@ -86,7 +87,7 @@ type PriorityPolicy struct {
 	Name string
 	// The numeric multiplier for the node scores that the priority function generates
 	// The weight should be a positive integer
-	Weight int
+	Weight int32
```
[the remark this answers is not in the document]
Comment (ahmad-diaa) [reply pkg/scheduler/api/types.go]: I used the the framework scoring step as a reference: https://github.com/kubernetes/kubernetes/blob/18559013ec0e8d3f4290b4c83386a404e148ec3b/pkg/scheduler/apis/config/types.go#L196-L201

Also, don't you think that int32 would be enough to represent a big range of numbers?

File: pkg/scheduler/api/types.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/api/types.go]: exporting because of a test is a bad reason, specially in an API package.

File: pkg/scheduler/api/types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/api/types.go]: I don't have a strong opinion on this. Let's see what other reviewers think.

File: pkg/scheduler/api/types.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (draveness) [reply pkg/scheduler/api/types.go]: `MaxWeight` could used by the user of kubernetes which could help them to limit the range of weight in extender or prioritiy functions.

File: pkg/scheduler/api/v1/types.go (hunk 1 above)
Comment (ahmad-diaa) [reply pkg/scheduler/api/v1/types.go]: Done

File: pkg/scheduler/algorithm/scheduler_interface.go [hunk 5]
```diff
@@ -39,7 +39,7 @@ type SchedulerExtender interface {
 	// Prioritize based on extender-implemented priority functions. The returned scores & weight
 	// are used to compute the weighted score for an extender. The weighted scores are added to
 	// the scores computed  by Kubernetes scheduler. The total scores are used to do the host selection.
-	Prioritize(pod *v1.Pod, nodes []*v1.Node) (hostPriorities *schedulerapi.HostPriorityList, weight int, err error)
+	Prioritize(pod *v1.Pod, nodes []*v1.Node) (hostPriorities *schedulerapi.HostPriorityList, weight int32, err error)
```
[the remark this answers is not in the document]
Comment (ahmad-diaa) [reply pkg/scheduler/algorithm/scheduler_interface.go]: Done.

File: pkg/scheduler/api/types.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ahmad-diaa) [reply pkg/scheduler/api/types.go]: @draveness unexporting MaxInt does not have anything to do with keeping MaxWeight Exported. Am I missing anything?

File: pkg/scheduler/algorithm/priorities/types.go [hunk 6]
```diff
@@ -50,7 +50,7 @@ type PriorityConfig struct {
 	// TODO: Remove it after migrating all functions to
 	// Map-Reduce pattern.
 	Function PriorityFunction
-	Weight   int
+	Weight   int32
```
Comment (alculquicondor) [inline pkg/scheduler/algorithm/priorities/types.go]: can we leave this one as int too?

File: pkg/scheduler/factory/plugins_test.go (hunk 2 above)
Comment (alculquicondor) [reply pkg/scheduler/factory/plugins_test.go]: There should still be a test for going over the limits.

File: pkg/scheduler/core/generic_scheduler.go [hunk 7]
```diff
@@ -1087,7 +1087,7 @@ func selectVictimsOnNode(
 	fitPredicates map[string]predicates.FitPredicate,
 	queue internalqueue.SchedulingQueue,
 	pdbs []*policy.PodDisruptionBudget,
-) ([]*v1.Pod, int, bool) {
+) ([]*v1.Pod, int32, bool) {
```
Comment (alculquicondor) [inline pkg/scheduler/core/generic_scheduler.go]: is this change necessary? I prefer we just change the API

Comment (ahmad-diaa) [reply pkg/scheduler/core/generic_scheduler.go]: It wouldn't be safe to return an int then cast it to int32
https://github.com/kubernetes/kubernetes/blob/84fe3db5cf58bf0fc8ff792b885465ceaf70a435/pkg/scheduler/core/generic_scheduler.go#L1014-L1021

should we change `schedulerapi.Victims.NumPDBViolations` to int64 then?

File: pkg/scheduler/factory/plugins_test.go (hunk 2 above)
Comment (ahmad-diaa) [reply pkg/scheduler/factory/plugins_test.go]: `PriorityConfig.Weight` (of type int32)*`schedulerapi.MaxPriority` (of value 10) wouldn't overflow in `int` type unless we have a very large number of PriorityConfigs in the test.

**Comment by k82cn** [comment]
@ahmad-diaa , would you help to send an email to api-machinery group on whether we should keep backward compatibility in code level? for example, the user have to update his code if he import those APIs.

**Comment by k82cn** [comment]
similar comments on `Lt/Gt` features

/cc @wgliang

File: pkg/scheduler/factory/plugins_test.go (hunk 2 above)
Comment (ahg-g) [reply pkg/scheduler/factory/plugins_test.go]: if you set MaxTotalPriority to math.MaxInt32, then those tests should still be valid, right?

File: pkg/scheduler/api/types.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ahg-g) [reply pkg/scheduler/api/types.go]: We should ask if we can change the values, I don't know whether or not this is considered part of the v1 API.

**Comment by ahg-g** [comment]
@liggitt can you please provide guidance here? we are trying to move to sized integers in this old API, we are wondering how we should do that given that this is already a v1 API, can we just change types? what about values of constants, are we allowed to change them?

**Comment by liggitt** [comment]
> @liggitt can you please provide guidance here? we are trying to move to sized integers in this old API, we are wondering how we should do that given that this is already a v1 API, can we just change types?

The important thing is to be able to continue deserializing config files you previously accepted. The safest way to do that is to switch unsized integers to int64.

json/yaml integers can't reliable serialize beyond 52 bits of precision (c.f. https://github.com/kubernetes/kubernetes/issues/17095), so someone using the highest possible int64 values could run into trouble with other json libraries, but that's less of a concern with config files consumed by a single binary than for REST APIs consumed by lots of clients implemented in lots of languages.

**Comment by liggitt** [comment]
The APIs we guarantee compatibility on are the REST APIs, config files, CLI flags, and persisted data. We do not guarantee go struct/interface/method compatibility release to release.

**Comment by ahg-g** [comment]
> The APIs we guarantee compatibility on are the REST APIs, config files, CLI flags, and persisted data. We do not guarantee go struct/interface/method compatibility release to release.
> 
> This PR is changing code inside the kubernetes/kubernetes repo that is not published for consumption as a go library, which is fine.
> 
> For repos we publish as libraries (things underneath the `staging` directory) [kubernetes/publishing-bot#196](https://github.com/kubernetes/publishing-bot/pull/196) is investigating adding support for publishing each kubernetes release under a new import path, so you could explicitly choose to switch to using (for example) `k8s.io/kube-scheduler/v16`

if I understand correctly, this implies that we are allowed to change (or even remove) the constants we define in https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/api/types.go#L27-L41 (those aren't even published under the versioned type: https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/api/v1/types.go).

File: pkg/scheduler/api/types.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/scheduler/api/types.go]: > We should ask if we can change the values, I don't know whether or not this is considered part of the v1 API.

Previously, a config file could specify a weight up to `9,223,372,036,854,775,807 / 10`, right? This change limits weights to `2,147,483,647 / 10`. Is there a reason not to switch to sized ints at int64 for compatibility?

File: pkg/scheduler/api/types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/scheduler/api/types.go]: unless there's a strong reason not to, keeping int64 for compatibility with all possible existing config files is preferable

**Comment by liggitt** [comment]
> if I understand correctly, this implies that we are allowed to change (or even remove) the constants we define in https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/api/types.go#L27-L41

You can remove the constants, but you should preserve the behavior and allowed values in existing config files (which could limit *changing* values)

File: pkg/scheduler/api/types.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ahg-g) [reply pkg/scheduler/api/types.go]: No, we can do that. I get your point now on what we should be preserving.

File: pkg/scheduler/core/generic_scheduler.go (hunk 7 above)
Comment (ahg-g) [reply pkg/scheduler/core/generic_scheduler.go]: changing NumPDBViolations to int64 will solve this issue.

File: pkg/scheduler/factory/plugins_test.go (hunk 2 above)
Comment (ahg-g) [reply pkg/scheduler/factory/plugins_test.go]: ditto, moving everything to int64 should address this as well .

File: pkg/scheduler/factory/plugins.go [hunk 8]
```diff
@@ -385,7 +385,7 @@ func RegisterCustomPriorityFunction(policy schedulerapi.PriorityPolicy) string {
 						policy.Argument.LabelPreference.Presence,
 					)
 				},
-				Weight: policy.Weight,
+				Weight: int(policy.Weight),
```
Comment (ahmad-diaa) [inline pkg/scheduler/factory/plugins.go]: > unless there's a strong reason not to, keeping int64 for compatibility with all possible existing config files is preferable

@ahg-g  @liggitt @alculquicondor  I think this would be a reason to have `PriorityPolicy.Weight` as int32 and not int64, otherwise we would change the type of `PriorityConfigFactory.Weight` to int64. What do you think?

**Comment by k82cn** [comment]
> The APIs we guarantee compatibility on are the REST APIs, config files, CLI flags, and persisted data. We do not guarantee go struct/interface/method compatibility release to release.

Great to know that, that'll be easier for us to refine the code :)

File: pkg/scheduler/algorithm/priorities/types.go [hunk 9]
```diff
@@ -50,7 +50,7 @@ type PriorityConfig struct {
 	// TODO: Remove it after migrating all functions to
 	// Map-Reduce pattern.
 	Function PriorityFunction
-	Weight   int
+	Weight   int64
```
[the remark this answers is not in the document]
Comment (ahmad-diaa) [reply pkg/scheduler/algorithm/priorities/types.go]: For doing so, we should keep MaxWeight as `MaxInt/MaxPriority` .. right?

[the remark this answers is not in the document]
Comment (ahmad-diaa) [reply pkg/scheduler/algorithm/priorities/types.go]: Also should the same check (to validate that the weight is less than MaxWeight) be added to ExtenderConfigs weight in 
https://github.com/kubernetes/kubernetes/blob/35cf6b6cbc7c55258ac0e3af51f6b0f5119750b1/pkg/scheduler/api/validation/validation.go#L45

like the one done in:
https://github.com/kubernetes/kubernetes/blob/35cf6b6cbc7c55258ac0e3af51f6b0f5119750b1/pkg/scheduler/api/validation/validation.go#L37

[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/algorithm/priorities/types.go]: I read that wrong. MaxWeight could still be big. I was thinking of Maxpriority.

File: pkg/scheduler/algorithm/priorities/reduce.go [hunk 10]
```diff
@@ -42,7 +42,7 @@ func NormalizeReduce(maxPriority int, reverse bool) PriorityReduceFunction {
 		if maxCount == 0 {
 			if reverse {
 				for i := range result {
-					result[i].Score = maxPriority
+					result[i].Score = int64(maxPriority)
```
[the remark this answers is not in the document]
Comment (ahmad-diaa) [reply pkg/scheduler/algorithm/priorities/reduce.go]: Do you mean changing the type passed to `NormalizeScore`?

File: pkg/scheduler/algorithm/priorities/types.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (ahmad-diaa) [reply pkg/scheduler/algorithm/priorities/types.go]: should it be kept as it is then?

File: pkg/scheduler/algorithm/priorities/reduce.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/algorithm/priorities/reduce.go]: yes
